### PR TITLE
Slightly optimize shroom spreading code

### PR DESCRIPTION
**Pull request by optimumtact**
There were some recent lag events on the servers caused by the timer subsystem jamming
up with glowshroom timers. I whined about this but nobody did anything so

There are a few changes here, first we collect the potential turfs for
the view call in a single pass, not once per yield run.

Second we dont' run a second view(1) on every single potential turf,
instead we try to randomly pick a turf 3 times during yield phase and
check the view then.

If we fail to find a potential location then we bail out of that yield
phase. This is a tradeoff between processing time spent finding
locations and the chance that the glowshroom fails to spread

Finally, we have maximum limit on how many times a glowshroom fails to
spread, if it fails to spread at least 5 times over any iteration, it
stops processing for spread completely.

As a bonus, the timers have been made unique, so we don't accidentally
generate multiple timers for a single shroom, other than the two it
already needs.

This code would benefit from being a seperate subsystem and grouping,
generations of the plants together as a single ticking entity and just
spreading from selected edge plants. However I don't particularly feel
like plumbing that together, so this will suit for now.

### CODE REVIEW

File: code/game/objects/effects/glowshroom.dm [hunk 1]
```diff
@@ -92,40 +95,60 @@
...
+
+			//Try three random locations to spawn before giving up tradeoff
+			//between running view(1, earth) on every single collected possibleLoc
+			//and failing to spread if we get 3 bad picks, which should only be a problem
+			//if there's a lot of glow shroom clustered about
+			for(var/Potato in 1 to 3)
+				var/turf/possibleLoc = pick(possibleLocs)
 				if(spreadsIntoAdjacent || !locate(/obj/structure/glowshroom) in view(1,earth))
-					possibleLocs += earth
-				CHECK_TICK
```
Comment (optimumtact) [inline code/game/objects/effects/glowshroom.dm:122]: You will note I have removed this.

If this causes server lag I will delete glowshrooms

Comment (LemonInTheDark) [reply code/game/objects/effects/glowshroom.dm:122]: based

File: code/game/objects/effects/glowshroom.dm [hunk 2]
```diff
@@ -92,39 +95,59 @@
...
 
 /obj/structure/glowshroom/proc/Spread()
+
+	//We could be deleted at any point and the timers might not be cleaned up
+	if(QDELETED(src))
+		return
+
 	var/turf/ownturf = get_turf(src)
 	var/shrooms_planted = 0
+	var/list/possibleLocs = list()
```
Comment (Timberpoes) [inline code/game/objects/effects/glowshroom.dm:114]: Can this be snake_case_ified, please?

File: code/game/objects/effects/glowshroom.dm [hunk 3]
```diff
@@ -92,39 +95,59 @@
...
-				CHECK_TICK
-
-			if(!possibleLocs.len)
-				break
 
-			var/turf/newLoc = pick(possibleLocs)
+		 // Whatever is the higher chance we use it (this is really stupid as the diminishing returns are effectively pointless???)
+		if(prob(max(chance_stats, chance_generation)))
+			var/spreadsIntoAdjacent = prob(spreadIntoAdjacentChance)
+			var/turf/newLoc = null
```
Comment (Timberpoes) [inline code/game/objects/effects/glowshroom.dm:136]: Can this be snake_case_ified, please?

File: code/game/objects/effects/glowshroom.dm [hunk 4]
```diff
@@ -92,39 +95,59 @@
...
+
+			//Try three random locations to spawn before giving up tradeoff
+			//between running view(1, earth) on every single collected possibleLoc
+			//and failing to spread if we get 3 bad picks, which should only be a problem
+			//if there's a lot of glow shroom clustered about
+			for(var/Potato in 1 to 3)
```
Comment (Timberpoes) [inline code/game/objects/effects/glowshroom.dm:142]: I can count to Potato.

**Review by Timberpoes — CHANGES_REQUESTED** [review]

File: code/game/objects/effects/glowshroom.dm (hunk 4 above)
Comment (LemonInTheDark) [reply code/game/objects/effects/glowshroom.dm:142]: is it a player ref if I don't play?

File: code/game/objects/effects/glowshroom.dm (hunk 3 above)
Comment (ExcessiveUseOfCobblestone) [reply code/game/objects/effects/glowshroom.dm:136]: you can also remove the `= null`

**Review by SpaceManiac — APPROVED** [review]
Code looks fine, improvements are reasonable, worked in testing

**Pull request merged**